apps/shell/tash_command.c

```diff
@@ -433,46 +435,67 @@ int check_exclam_cmd(char *buff)
 
 	/* Find the '!' in the input character */
 
-	while ((exclam_ptr = strchr(buff, ASCII_EXCLAM)) != NULL) {
+	int qc = NOT_IN_QUOTES;		/* Flag to check whether ! is inside quotation mark */
 
-		exclam_nextptr = exclam_ptr + 1;
-		if (*exclam_nextptr == (char)ASCII_SPACE || *exclam_nextptr == (char)ASCII_LF || *exclam_nextptr == (char)ASCII_SEMICOLON || *exclam_nextptr == (char)ASCII_NUL) {
-			/* If the last character in buff is '!', it will be handled as character, not history command. */
+	while (*buff != (char)ASCII_NUL) {
 
-			buff = exclam_nextptr + 1;
+		if(*buff == (char)ASCII_RSQUOTE && qc == NOT_IN_QUOTES) {
```

Comment:
Please find my previous comment.
```
if ( ) {
```
You need to check TizenRT coding rules first.